### ISSUE TITLE
Should Nix self-host the binary cache?

### ISSUE THREAD

**Issue by nyabinary** [NixOS/SC-election-2024#19]
### Question

Do you believe that Nix should transition to self-hosting the binary cache on bare-metal hardware, rather than relying on third-party services for hosting its cache?

### Reminder of the Q&A rules

Please adhere to the [Q&A guidelines and rules](https://github.com/NixOS/SC-election-2024/blob/main/doc/qna.md)

**Comment by cafkafk** [comment]
TL;DR:
- Yes, but it will take a while, and cost a lot.
- Until we get there, we should try to be as agnostic towards specific providers as possible.
- We should provide caching to both official and community projects.

To the extend possible, yes. Currently, data center space is expensive, and at our scale, we're likely no longer looking at a hobbyist garage full of towers. We have to accept that for a time, we'll likely not have the option to say no to cloud providers, and that they can be useful tools, as they often provide rapid scaling and can be useful for disaster recovery.

That said, we should ensure that infrastructure is made in a cloud agnostic way, so as to be as independent as possible from a single cloud provider, and to make the shift to a self hosted solution as pain free as possible. This also means restricting usage of proprietary software in our infrastructure, to ensure its longevity.

At the end of the day, we can't control when we'll be ready to make the move financially, but we can be as prepared as reasonably possible for when it becomes an option. That said, I think the costs is too prohibitive for our current financial position, but it's very reasonable that it will be possible in the future.

Another thing worth mentioning is that we should consider providing cache to more projects. Ideally, I think the foundation should host a community cache, that would allow tenancy of projects compatible with our values, and meeting certain criteria such as being of a certain size. Making cache more available to smaller projects would be a way in which we can help foster a more lively ecosystem, and user our resources in a way that is useful to the community at large.

**Comment by mschwaig** [comment]
Long term, I think it makes sense to invest in a peer-to-peer system to distribute traffic and storage, since that fits really well with how Nix works conceptually.

In the short term, it's primarily a matter of cost and risk to me.
I think some people on discourse already did some calculations. I would listen to what they say and see who is willing to put in the work.

**Comment by tomberek** [comment]
I don't know. Conceptually, I like the idea of a community-maintained peer-to-peer cache and self-hosting. Self-hosting requires even more work to be placed onto a limited infrastructure team and either appproach requires additional development to be a viable alternative. Thus I think the right approach is to allow any of these to be experimented with:

> TIP: When there is an interesting problem, try to get multiple teams competing to solve it. Competition is great fun and can produce better answers than monopolized problems. You can even explicitly create competitions with prizes for the best solutions.
    - https://hintjens.gitbooks.io/social-architecture/content/chapter1.html

**Comment by getchoo** [comment]
Unless there is a clear and concise path for doing this long-term (i.e., funding, partnership, etc), no. The community and Foundation have already had [issues](https://discourse.nixos.org/t/the-nixos-foundations-call-to-action-s3-costs-require-community-support/28672) in maintaining the binary cache *with* outside support. Moving this completely in-house would most likely exaggerate this issue without a partner, and would definitely affect people in areas not in the immediate proximity of the bare metal hardware if we were to drop Fastly as well

As mentioned by others here, some [nice work](https://github.com/nix-community/trustix) has also gone towards making a P2P binary cache viable, which could very well take the burden off the Foundation/NixOS org and help with the CDN issue. I find this to be more of something to consider in the (somewhat) far future though, as currently they are not production ready and haven't been tested at the scale of our current binary cache

**Comment by nyabinary** [comment]
I agree with Cafkafk's approach on transitioning to self-hosting the binary cache, but I also believe we should focus on immediate steps to manage costs, such as garbage collecting the current cache to keep costs low(er). While we work toward a self-hosted solution, maintaining provider-agnosticity is key, and careful management of existing resources like the cache will help minimize financial strain. In the long term, transitioning to a self-hosted, cloud-agnostic solution will strengthen Nix's independence and better align with our principles.

**Comment by roberth** [comment]
Self hosting may be a good option, when we have a good plan. Funding a self-hosted replacement for the S3 part of the solution _shouldn't_ be a problem, considering our large user base and perhaps even the EU's "digital sovereignty" goals.

As far as I know, we're still looking into multiple options, so I see no reason to take self-hosting off the table, but it does need a good plan.

**Comment by winterqt** [comment]
I don't have much to add, but I agree with @yu-re-ka and @roberth's take. It's doable, but takes planning and good execution to be done well. I know we'll be able to do it eventually with time, though.

**Comment by proofconstruction** [comment]
As answered in #16, I would like to eventually achieve full reproducibility and ultimately remove the need for significant caching infrastructure, but in the meantime we should consider alternatives to the present situation. Self-hosting comes with considerable challenges and would require significantly more infrastructure support which we presently have neither the labor power nor the financial capacity to address.
